#!/usr/bin/env python2.6

#
# Exit codes:
#   0: OK
#   1: Miscellaneous error
#   2: Connection refused
#   3: Authentication failed
#   4: xapi password unconfigured
#

import socket
import sys
import xml.parsers.expat

import XenAPI

import os_vpx


try:
    session = os_vpx.login()
    # Check for OS-356.  10 iterations seemed to be enough generally -- 50
    # is deliberate overkill.
    for i in xrange(0, 50):
        session.xenapi.VM.get_all_records()
except xml.parsers.expat.ExpatError, exn:
    print >>sys.stderr, 'Error: OS-356.'
    sys.exit(5)
except os_vpx.XapiPasswordUnconfigured:
    print >>sys.stderr, 'Error: xapi password unconfigured.'
    sys.exit(4)
except XenAPI.Failure, exn:
    if exn.details[0] == 'SESSION_AUTHENTICATION_FAILED':
        print >>sys.stderr, 'Error: authentication to xapi failed.'
        sys.exit(3)
    else:
        print >>sys.stderr, exn
        sys.exit(1)
except socket.error, exn:
    if exn.errno == 111:
        print >>sys.stderr, \
              'Error: connection to %s refused.' % os_vpx.connection_url()
        sys.exit(2)
    else:
        print >>sys.stderr, exn
        sys.exit(1)
